### DISCUSSION TITLE
Training with Graph Mini-Batches

### DISCUSSION THREAD

**Discussion by kerem0comert** [pyg-team/pytorch_geometric#6355] [Q&A] [answered]
Hello,

I would like to perform graph-classification where each graph is of the following format:
Data(x=[21, 14], edge_index=[2, 42]), where each graph has 21 nodes with each node having 14 features, and 21 edges that connect them.

As such, my whole dataset has the following format where I have n=108221 such graphs:
graph_batch = DataBatch(x=[2272620, 14], edge_index=[2, 4545240], batch=[2272620], ptr=[108221])

Given the size of my dataset, the code which I used previously for smaller tasks no longer work on my single RTX3070 GPU, as I run out of memory.

As such, I would like to train with [PyTorch Mini-Batches](https://pytorch-geometric.readthedocs.io/en/latest/notes/batching.html), but wanted to ask the community for your opinions to see how I can achieve it using best practices.

Before implementing Mini-Batches, my working code was the following (although in bigger datasets I do not have enough memory to run it):
```python
%%time
device = torch.device('cuda' if torch.cuda.is_available() else 'cpu')
model = GATConvModel(in_features=graph_batch.x.shape[1], hidden_features=8, num_classes=len(classes), pool_type="mean").to(device)
data = graph_batch.to(device)

train_percentage = 0.8

train_count = ceil(train_percentage * graph_batch.num_graphs)
test_count = graph_batch.num_graphs - train_count
data.train_mask = torch.tensor([True] * train_count + [False] * test_count)

# Randomly shuffle the train_mask
data.train_mask = data.train_mask[torch.randperm(data.train_mask.size(0))]
data.test_mask = ~data.train_mask
data.y = torch.tensor(y_dataset).type(torch.LongTensor)
optimizer = torch.optim.Adam(model.parameters(), lr=0.005, weight_decay=5e-4)
model.train()

losses = []
for epoch in range(1000):
    optimizer.zero_grad()
    out = model(data.x, data.edge_index, data.batch)
    loss = F.nll_loss(out[data.train_mask], data.y[data.train_mask])
    
    if epoch%10 == 0:
        print(f"Epoch {epoch}, Loss: {loss.item():.4f}")
        losses.append(loss.item())
    
    loss.backward()
    optimizer.step()

# Plot the loss over the epochs
plt.plot(losses)
plt.show()
```

How can I fix the following code to achieve what I am looking for?
```python
%%time

device = torch.device('cuda' if torch.cuda.is_available() and args.use_cuda else 'cpu')
model = GATConvModel(in_features=graph_batch.x.shape[1], 
                     hidden_features=8, num_classes=len(classes), pool_type="mean").to(device)

train_percentage = 0.8

train_count = ceil(train_percentage * graph_batch.num_graphs)
test_count = graph_batch.num_graphs - train_count
graph_batch.train_mask = torch.tensor([True] * train_count + [False] * test_count)

# Randomly shuffle the train_mask
graph_batch.train_mask = graph_batch.train_mask[torch.randperm(graph_batch.train_mask.size(0))]
graph_batch.test_mask = ~graph_batch.train_mask
graph_batch.y = torch.tensor(y_dataset).type(torch.LongTensor)
graph_batch.to(device)

dataloader = DataLoader(graph_batch, batch_size=args.batch_size)

optimizer = torch.optim.Adam(model.parameters(), lr=0.005, weight_decay=5e-4)
model.train()

losses = []
for epoch in range(1000):
    optimizer.zero_grad()
    for data in dataloader:
        out = model(data.x, data.edge_index, data.batch)
        
        # Here what should the line look like?
        loss = F.nll_loss(out[graph_batch.train_mask], graph_batch.y[graph_batch.train_mask])
        
        if epoch%10 == 0:
            print(f"Epoch {epoch}, Loss: {loss.item():.4f}")
            losses.append(loss.item())
            loss.backward()
    optimizer.step()

# Plot the loss over the epochs
plt.plot(losses)
plt.show()
```

Thanks in advance for your inputs :)

**Comment by wsad1** [comment, marked as answer]
Firstly don't combine your list of `Data` objects into a `Data` object. `DataLoader` can work with a list of `Data` objects.
```
graph_list: List[Data] = ... # list of `Data` objects
train_loader = DataLoader(graph_list[train_index], batch_size=args.batch_size)
test_loader = DataLoader(graph_list[test_index], batch_size=args.batch_size)
```
You can directly use these data loaders in the above code. 
Refer to graph classifications examples in pyg for more information. https://github.com/pyg-team/pytorch_geometric/blob/master/examples/proteins_topk_pool.py, https://github.com/pyg-team/pytorch_geometric/blob/master/examples/mem_pool.py etc.

**Reply by kerem0comert** [reply]
Thank you! Having made the changes that you specified by also following the first example that you shared specifically, I now have a version that uses List[Data] objects rather than one giant disconnected graph as was the case with DataBatch.
However, training in this way without batching at all, I realized that the training time is significantly higher and the accuracy is much worse.

For instance, one of the datasets I train on has the following properties:
```
Total number of samples = (926,) where each sample is Data(x=[21, 14], edge_index=[2, 42])
```

- When I load it in the DataBatch way so that it looks like this:
```python
DataBatch(x=[19446, 14], edge_index=[2, 38892], batch=[19446], ptr=[927]) # Note that t19446 / 926 = 21
```
My results are the following with 0.98 accuracy:
![databatch](https://user-images.githubusercontent.com/42437138/212690426-6ff19777-5ebd-4846-97dd-7428075c6cd4.jpg)

- When I load it as List[Data] without mini-batching (so that batch_size = num_of_graphs) the graph_list is a list of 926 Data(x=[21, 14], edge_index=[2, 42]) items.
In this case however, my results are as follows with 0.43 accuracy:
![data](https://user-images.githubusercontent.com/42437138/212690823-c4c668ba-0aaa-4796-b180-0bbf0c6a522c.jpg)

I was not able to pinpoint what goes wrong exactly, so I am pasting my code snippets for the second training I run. Would be happy to hear your opinions and thanks in advance.

## Loading the data

```python
with open(LABELS_FILE_PATH,encoding='utf-8-sig') as f:
    classes = [row[0] for row in csv.reader(f)]
    print(f"{classes=}")
    NUM_CLASSES = len(classes)
y = np.loadtxt(os.path.join(os.getcwd(), DATASET_FILE_PATH), delimiter=',', dtype='int32', usecols=(0))

gesture_recognizer = GestureRecognizer(args=args, init_stream=False)
input_shape = gesture_recognizer.get_input_shape()
graph_list: list[Data] = gesture_recognizer.get_gnn_batch(use_one_giant_graph=False, 
                                                          dataset_file_path=os.path.join(os.getcwd(), DATASET_FILE_PATH)
                                                          )
device = torch.device('cuda' if torch.cuda.is_available() and args.use_cuda else 'cpu')

print(graph_list[0]) # Print the first graph in the batch
print(len(graph_list))
print(f"Total number of samples = {y.shape}")

```

```python
BATCH_SIZE = y.shape[0] // args.batch_scaler # In this case args.batch_scaler = 1 so there is no mini-batching.

train_count = ceil(args.train_percentage * y.shape[0])
test_count = y.shape[0] - train_count
train_mask = torch.tensor([True] * train_count + [False] * test_count)

# Randomly shuffle the train_mask
train_mask = train_mask[torch.randperm(train_mask.size(0))]
test_mask = ~train_mask

# From the graph_batch which is a List[Data], only get the items whose index is a value of True in the train_mask
train_loader = DataLoader([graph_list[i] for i in range(len(graph_list)) if train_mask[i]])
test_loader = DataLoader([graph_list[i] for i in range(len(graph_list)) if test_mask[i]])

train_y = torch.tensor([y[i] for i in range(len(y)) if train_mask[i]], dtype=torch.int64, device=device)
# train_y = train_y.reshape(train_y.shape[0] // BATCH_SIZE if BATCH_SIZE, BATCH_SIZE)
test_y = torch.tensor([y[i] for i in range(len(y)) if test_mask[i]], dtype=torch.int64, device=device)
# test_y = test_y.reshape(test_y.shape[0] // BATCH_SIZE, BATCH_SIZE)

print(f"Example graph = {train_loader.dataset[0]}")
print(f"{len(train_loader.dataset)} training samples with batch size={BATCH_SIZE}")
print(f"{len(test_loader.dataset)} test samples with batch size={BATCH_SIZE}")
print(train_y.shape)
```

## Train/test functions
```python

def train(epoch, y, model, device, optimizer, train_dataset):
    model.train()

    loss_all = 0
    for i, data in enumerate(train_loader):
        data = data.to(device)
        optimizer.zero_grad()
        output = model(data.x, data.edge_index, data.batch)
        # Reshape y[i] from torch.Size([]) to torch.Size([1])
        loss = F.nll_loss(output, y[i].reshape(1))
        loss.backward()
        loss_all += data.num_graphs * loss.item()
        print(f"{epoch}, Batch: {i}, Loss: {loss.item()}")
        optimizer.step()
    return loss_all / len(train_dataset)

def test(loader, y, model, device):
    model.eval()

    correct = 0
    predictions = []
    for i, data in enumerate(loader):
        data = data.to(device)
        pred = model(data.x, data.edge_index, data.batch).max(dim=1)[1]
        predictions.append(pred)
        correct += pred.eq(y[i].reshape(1)).sum().item()
    return correct / len(loader.dataset), torch.as_tensor(predictions)
```

```python
%%time

## Training
model = GATConvModel(in_features=graph_list[0].x.shape[1], 
                     hidden_features=8, num_classes=len(classes), pool_type="mean").to(device)

optimizer = torch.optim.Adam(model.parameters(), lr=args.learning_rate, weight_decay=args.weight_decay)
model.train()

losses = []
for epoch in range(args.epochs):
    start_time: float = timer()
    losses.append(train(epoch, train_y, model, device, optimizer, train_loader.dataset))
    train_acc, _ = test(train_loader, train_y, model, device)
    if len(test_y):
        test_acc, _ = test(test_loader, test_y, model, device)
    else: test_acc = 0
    # print(f'Epoch: {epoch:03d}, Loss: {losses[-1]}, Train Acc: {train_acc:.5f}, '
    #      f'Test Acc: {test_acc:.5f}. Epoch time: {timer() - start_time:.2f}s')

# Plot the loss over the epochs
plt.plot(losses)
plt.show()
```

**Reply by wsad1** [reply]
Are you using the pyg [DataLoader](https://pytorch-geometric.readthedocs.io/en/latest/modules/loader.html#torch_geometric.loader.DataLoader)? Also you'll have to train with a larger `batch_size` for it to converge to a good solution faster.

**Reply by kerem0comert** [reply]
Yes, I am using pyg's DataLoader. 
Definitely, my plan is to use with bigger batch sizes eventually. But just as a baseline, I first wanted to work within the framework of batching but without any batching initially. 
I can maybe understand the training time to be longer (given that in the first case I use one giant graph whereas in the second came each time graph data gets loaded and unloaded perhaps? So maybe a space vs time complexity problem), but I was not expecting to see such a significant disparity.

Also in any case, I am not sure how to justify the weird loss curve and such a decrease in accuracy, so clearly I am doing something wrong.

**Reply by wsad1** [reply]
I am not entirely sure whats going wrong. In the code snippet you pasted can you try setting `batch_size` to `926`, and then train the model.

**Reply by kerem0comert** [reply]
With manually setting the BATCH_SIZE, same issue. Thank you nevertheless for your time though, perhaps someone else could also shed some light into this.

**Reply by rusty1s** [reply]
Do you have a TLDR of the issue here? We provide many examples for learning in a mini-batch scenario, e.g., `examples/ppi` or `examples/mutag_gin`. Note that you also want to use `shuffle=True` in the `DataLoader` for training examples.

**Reply by kerem0comert** [reply]
I have modeled my script after your examples, and the code does run - the issue is about the result.
Regarding shuffle, I am aware that it exists in ```DataLoader```, but for now I used my own logic to shuffle that works, albeit it is much more verbose (the reason is that in my first approach I do not use the ```DataLoader```.

TLDR: I have a graph-classification problem. Each graph has 21 nodes, 21 directed edges with 14 features for each node (although 12 of these features are the same and repeated across all nodes)
In my first approach, I load the data in one giant graph, where each of my actual graphs are disconnected components in the giant graph.
As such, for a given dataset with n=108221 graphs, I have the following:
```python
# Note that x = 2272620 ≅ 21 * n
graph_batch = DataBatch(x=[2272620, 14], edge_index=[2, 4545240], batch=[2272620], ptr=[108221])
```
This works well in smaller datasets, but I run out of VRAM in bigger datasets like the one presented here.

As such I would like to use mini-batching. As a first step, I no longer load one giant graph and rather load the graphs in a list as @wsad1 suggested, so the same dataset is the following:,
```python
graph_list is a list[Data] of length=108022 where each item in the list is Data(x=[21, 14], edge_index=[2, 42])
```
When I train this way using the code snippets [here](https://github.com/pyg-team/pytorch_geometric/discussions/6355#discussioncomment-4697452), I get much worse accuracy, training time and loss curves. This is irregardless of if I set the batch_size to a specific value or I set it to be equal to n (effectively not performing mini-batching at all).

I am looking forward to your ideas as to why this could be the case.

**Reply by rusty1s** [reply]
This is weird. Do you achieve original accuracy when running with `batch_size=len(dataset)`?

**Reply by kerem0comert** [reply]
No, I don't unfortunately. In fact this is the first thing I tried, this is what I mean by saying:
```This is irregardless of if I set the batch_size to a specific value or I set it to be equal to n (effectively not performing mini-batching at all).```

**Reply by rusty1s** [reply]
But how did you create `DataBatch` on the full dataset in the first place? I guess you are using `Batch.from_data_list`, right? This is exactly what `DataLoader` is doing internally as well, so I am very confused. If you run
```python
loader = DataLoader(dataset, batch_size=len(dataset))
batch = next(iter(loader))
```
this should yield exactly the same `graph_batch` as you created on your own. Can you confirm?

**Reply by kerem0comert** [reply]
Yes, you are correct—I am using ```Batch.from_data_list```. 
You are also right that if I do:
```python
   graphs = [tg.data.Data(x=torch.tensor(x), edge_index=EDGE_INDEX) for x in X_dataset]
   return graphs
```
rather than 
```python
graphs = [tg.data.Data(x=torch.tensor(x), edge_index=EDGE_INDEX) for x in X_dataset]
return tg.data.batch.Batch.from_data_list(graphs)
```
then I can get the same graph_batch by doing:
```python
loader = DataLoader(dataset, batch_size=len(dataset))
batch = next(iter(loader))
```

**Reply by rusty1s** [reply]
Super, does that mean that you found the issue?

**Reply by kerem0comert** [reply]
Unfortunately not. The problem is not that the code does not compile, both versions do and produce a result in fact. It is just that I would like to move my implementation from a giant ```graph_batch``` to a ```List[Data]``` object so that I can mini-batch in bigger datasets (the one giant ```graph_batch``` for a dataset with 100k samples run out of VRAM for instance).

But as can be seen from [this](https://github.com/pyg-team/pytorch_geometric/discussions/6355#discussioncomment-4697452) answer, the training I do with ```List[Data]``` does compile, but produces significantly worse accuracy, training time and loss curve.
I just wanted to find out what would be the possible cause of this, and whether I am doing something wrong in my implementation to produce this bad result despite the code compiling.

**Reply by rusty1s** [reply]
This is still confusing, I am so sorry :D 

In the answer you linked you said that the performance differs between one giant graph and `Dataloader(..., batch_size=len(dataset)` (0.98 vs 0.43), but you now get the same graphs back. So are you now able to get to 0.98 when using `batch_size=len(dataset)` when using `DataLoader`?

**Reply by kerem0comert** [reply]
Totally fine, probably I explained it rather confusingly - thanks for helping out :)
I do get the same graph back, if I use ```batch = next(iter(loader))``` on it. The problem is, training with the giant graph is only possible to a certain extent, because after some point with bigger datasets, I run out of VRAM. This is why I wanted to explore mini-batching at the first place, and in order to mini-batch I thought that I need to use the DataLoader first, and load with ```List[Data]``` as @wsad1 suggested.

So as it currently stands, my dilemma is as follows:
- Training with the giant graph yields perfect accuracy and perfect training time but it is not possible to do when the dataset gets bigger, as I am bound by the VRAM that I have (currently it is 8GB. I can have access to 24GB and train throughout the week if necessary, but at the end of the day this approach may not be scalable)
- Training with ```DataLoader``` using ```List[Data]``` seems to be working as I do not run out of VRAM, but the training time is significantly longer and more crucially the accuracy is worse to the point of the model not being useful.

So as I see it - either,
- I should mini-batch with the giant graph so that the data would fit to my VRAM
- Or I should fix the ```DataLoader``` results somehow and use that implementation.

I would be happy to hear your thoughts on either approaches, as so far I got stuck :/

**Reply by rusty1s** [reply]
Yes, that's what I understand. Just wanted to make sure you reach the same (good) performance on the full dataset independent of whether you collate by yourself or via `DataLoader`.

I personally have never seen a drop in performance when using mini-batches, in most cases, it works even better due to its stochastic nature. Overall, it should also be faster as you do more optimization steps. Importantly, there might be some hyperparameters you want to tune when running in mini-batches.
* `batch_size` shouldn't be too small. How is it currently set?
* You are now doing way more optimization steps, so you may need to tune the `learning_rate` to account for this.

Looking at your code, I am not totally sure why you only perform optimization steps a single time per epoch (and only backward every 10th epoch). It should be
```python
for epoch in range(1000):
    for data in dataloader:
        optimizer.zero_grad()
        out = model(data.x, data.edge_index, data.batch)
        
        # Here what should the line look like?
        loss = F.nll_loss(out[graph_batch.train_mask], graph_batch.y[graph_batch.train_mask])
        
        loss.backward()
        optimizer.step()
```

As a rule of thumb, you probably don't want to deal with creating mini-batches by yourself in PyG - that's the task of `DataLoader`.

**Reply by kerem0comert** [reply]
After much tinkering with the code, it seems to be working smoothly now, also in the case of ```DataLoader```. I was not sure of the exact root cause, but probably it was a combination of factors where I was not sending parameters correctly to the PyG methods. Thank you so much @rusty1s  @wsad1 for your kind answers, they were really helpful to nudge me into the right direction :)

For future reference to other potential seekers, I am pasting the final code which produces a reasonable result:

## Loading Data
```python
BATCH_SIZE = y.shape[0] // args.batch_scaler

train_count = ceil(args.train_percentage * y.shape[0])
test_count = y.shape[0] - train_count
train_mask = torch.tensor([True] * train_count + [False] * test_count)

# Randomly shuffle the train_mask
train_mask = train_mask[torch.randperm(train_mask.size(0))]
test_mask = ~train_mask

train_loader = DataLoader([graph_list[i] for i in range(len(graph_list)) if train_mask[i]], batch_size=BATCH_SIZE)
test_loader = DataLoader([graph_list[i] for i in range(len(graph_list)) if test_mask[i]], batch_size=BATCH_SIZE)

# train_y = train_y.reshape(train_y.shape[0] // BATCH_SIZE, BATCH_SIZE)

# test_y = test_y.reshape(test_y.shape[0] // BATCH_SIZE, BATCH_SIZE)

print(f"Example graph = {train_loader.dataset[0]}")
print(f"{len(train_loader.dataset)} training samples with batch size={BATCH_SIZE}")
print(f"{len(test_loader.dataset)} test samples with batch size={BATCH_SIZE}")
```

## train/test methods
```python
def train(epoch, loader, model, device, optimizer):
    model.train()
    loss_all = 0
    for i, data in enumerate(loader):
        data = data.to(device)
        optimizer.zero_grad()
        output = model(data.x, data.edge_index, data.batch)
        loss = F.nll_loss(output, data.y)
        loss.backward()
        loss_all += loss.item() * data.num_graphs
        optimizer.step()
    return loss_all / len(loader.dataset)

def test(loader, model, device):
    model.eval()

    correct = 0
    predictions = []
    for i, data in enumerate(loader):
        data = data.to(device)
        pred = model(data.x, data.edge_index, data.batch).max(dim=1)[1]
        predictions.append(pred)
        correct += pred.eq(data.y).sum().item()
    return correct / len(loader.dataset), predictions
```

## Training
```python
%%time


model = GATConvModel(in_features=graph_list[0].x.shape[1], 
                     hidden_features=8, num_classes=len(classes), pool_type="mean").to(device)

optimizer = torch.optim.Adam(model.parameters(), lr=args.learning_rate, weight_decay=args.weight_decay)
model.train()

losses = []
for epoch in range(args.epochs):
    start_time: float = timer()
    losses.append(train(epoch, train_loader, model, device, optimizer))
    train_acc, _ = test(train_loader, model, device)
    if args.train_percentage < 1:
        test_acc, _ = test(test_loader, model, device)
    else:
        test_acc = train_acc
    print(f'Epoch: {epoch:03d}, Loss: {losses[-1]}, Train Acc: {train_acc:.5f}, '
          f'Test Acc: {test_acc:.5f}. Epoch time: {timer() - start_time:.2f}s')

# Plot the loss over the epochs
plt.plot(losses)
plt.show()
```

## Results
```python
test_loader = DataLoader([graph_list[i] for i in range(len(graph_list)) if test_mask[i]], batch_size=1)
accuracy, predictions = test(test_loader, model, device)
print_confusion_matrix(classes=classes,
                       y_true=y[test_mask], 
                       y_pred=torch.Tensor.cpu(torch.tensor(predictions))
)
```
